gateway/tests/reader_test.go

```diff
@@ -0,0 +1,208 @@
+package tests
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/openfaas/faas/gateway/handlers"
+	"github.com/openfaas/faas/gateway/metrics"
+	"github.com/openfaas/faas/gateway/requests"
+	"golang.org/x/net/context"
+)
+
+type testServiceApiClient struct {
+	serviceListServices []swarm.Service
+	serviceListError    error
+}
+
+func (c testServiceApiClient) ServiceCreate(ctx context.Context, service swarm.ServiceSpec, options types.ServiceCreateOptions) (types.ServiceCreateResponse, error) {
+	return types.ServiceCreateResponse{}, nil
+}
+
+func (c testServiceApiClient) ServiceInspectWithRaw(ctx context.Context, serviceID string, options types.ServiceInspectOptions) (swarm.Service, []byte, error) {
+	return swarm.Service{}, []byte{}, nil
+}
+
+func (c testServiceApiClient) ServiceList(ctx context.Context, options types.ServiceListOptions) ([]swarm.Service, error) {
+	return c.serviceListServices, c.serviceListError
+}
+
+func (c testServiceApiClient) ServiceRemove(ctx context.Context, serviceID string) error {
+	return nil
+}
+
+func (c testServiceApiClient) ServiceUpdate(ctx context.Context, serviceID string, version swarm.Version, service swarm.ServiceSpec, options types.ServiceUpdateOptions) (types.ServiceUpdateResponse, error) {
+	return types.ServiceUpdateResponse{}, nil
+}
+
+func (c testServiceApiClient) ServiceLogs(ctx context.Context, serviceID string, options types.ContainerLogsOptions) (io.ReadCloser, error) {
+	return nil, nil
+}
+
+func (c testServiceApiClient) TaskLogs(ctx context.Context, taskID string, options types.ContainerLogsOptions) (io.ReadCloser, error) {
+	return nil, nil
+}
+
+func (c testServiceApiClient) TaskInspectWithRaw(ctx context.Context, taskID string) (swarm.Task, []byte, error) {
+	return swarm.Task{}, []byte{}, nil
+}
+
+func (c testServiceApiClient) TaskList(ctx context.Context, options types.TaskListOptions) ([]swarm.Task, error) {
+	return []swarm.Task{}, nil
+}
```

Comment:
It's kind of a shame to have so many methods here including Remove/Create etc. Might be necessary for using the interface verbatim. the adapter pattern could allow for a smaller interface implementation.

Reply:
That's true. I wanted to avoid changing `server.go` so I went with the approach of using the interface from docker. Do you want me to create an adapter for this PR or address it in the following REST-related PRs?

Reply:
I'll merge this as-is, but yes It think we should be careful of having this sort of repetition.